ramenpos/RamenPos/Logic/RamenApp+SPIConnectionDelegate.swift

```diff
@@ -0,0 +1,91 @@
+//
+//  SPIConnectionDelegate.swift
+//  RamenPos
+//
+//  Created by Amir Kamali on 28/5/18.
+//  Copyright © 2018 Assembly Payments. All rights reserved.
+//
+
+import Foundation
+import SPIClient_iOS
+
+extension RamenApp: SPIDelegate {
+    
+    /// Called when we received a Status Update i.e. Unpaired/PairedConnecting/PairedConnected
+    func spi(_ spi: SPIClient, statusChanged state: SPIState) {
+        NotificationCenter.default.post(name: NSNotification.Name(rawValue: AppEvent.connectionStatusChanged.rawValue), object: state)
+        
+        SPILogMsg("statusChanged \(state)")
+    }
+    
+    // Called during the pairing process to let us know how it's going.
+    // We just update our screen with the information, and provide relevant Actions to the user.
+    func spi(_ spi: SPIClient, pairingFlowStateChanged state: SPIState) {
+        SPILogMsg("pairingFlowStateChanged \(state)")
+        
+        NotificationCenter.default.post(name: NSNotification.Name(rawValue: AppEvent.pairingFlowChanged.rawValue), object: state)
+    }
+    
+    func spi(_ spi: SPIClient!, secretsChanged secrets: SPISecrets?, state: SPIState!) {
+        SPILogMsg("secrets \(String(describing: state))")
+        
+        if let secrets = secrets {
+            SPILogMsg("\n\n")
+            SPILogMsg("# ----------- I GOT NEW SECRETS -----------")
+            SPILogMsg("# ---------- PERSIST THEM SAFELY ----------")
+            SPILogMsg("# \(String(describing: secrets.encKey)):\(String(describing: secrets.hmacKey))")
+            SPILogMsg("# -----------------------------------------")
+            
+            // In prod store them in the key chain
+            settings.encriptionKey = secrets.encKey!
```

Comment:
Probably best to start off a new project with as few `!` as possible, they quickly get out of control if you let them, and you can end up crashier than the equivalent objc. If let or guard are good options.

Reply:
Removed '!'